src/sentry/api/endpoints/grouping_level_details.py

```diff
@@ -0,0 +1,194 @@
+        destination_id=None,
+        fingerprints=None,
+        actor_id=request.user.id if request.user else None,
+        replacement=replacement,
+    )
+
+
+def _decrease_level(group: Group, id: int, levels_overview: LevelsOverview, request):
+    assert levels_overview.parent_hashes
+    assert id < len(levels_overview.parent_hashes)
+
+    new_materialized_hash = levels_overview.parent_hashes[id]
+
+    # Create new group with parent hash that we want to group by going forward.
+    destination = Group.objects.create(
+        project_id=group.project_id, short_id=group.project.next_short_id()
+    )
+
+    GroupHash.objects.create_or_update(
+        project_id=group.project_id,
+        hash=new_materialized_hash,
+        defaults={"group_id": destination.id},
+    )
+
+    now = datetime.datetime.now()
+
+    # When decreasing the level, events from this + other groups get merged
+    # into one new parent group. This is why we cannot restrict the query by
+    # timestamp here. The query logic is analoguous to
+    # GroupingLevelNewIssuesEndpoint.
+
+    query = (
+        Query("events", Entity("events"))
+        .set_select(
+            [
+                Column("group_id"),
+                Function(
+                    "argMax",
+                    [
+                        Function(
+                            "arraySlice",
+                            [
+                                Column("hierarchical_hashes"),
+                                id + 1,
+                            ],
+                        ),
+                        Column("timestamp"),
+                    ],
+                    "reset_hashes",
+                ),
+            ]
+        )
+        .set_groupby([Column("group_id")])
+        .set_where(
+            [
+                Condition(Column("primary_hash"), Op.EQ, levels_overview.only_primary_hash),
+                Condition(Column("project_id"), Op.EQ, group.project_id),
+                Condition(
+                    Function("arrayElement", [Column("hierarchical_hashes"), id + 1]),
+                    Op.EQ,
+                    new_materialized_hash,
+                ),
+                Condition(Column("timestamp"), Op.GTE, now - datetime.timedelta(days=90)),
+                Condition(Column("timestamp"), Op.LT, now + datetime.timedelta(seconds=10)),
+            ]
+        )
+    )
+
+    # TODO pagination
+    # We expect this query to return <200 results, so this should still be
+    # acceptable especially considering the loop body is fast.
+    #
+    # If we move this into a paginating celery task we need to consider that
+    # we're iterating over a table that is also modified by the unmerge task at
+    # the same time.
+    for row in snuba.raw_snql_query(
+        query, referrer="api.grouping_levels_details.decrease_level.get_group_ids"
+    )["data"]:
+        group_id = row["group_id"]
+
+        replacement = HierarchicalUnmergeReplacement(
```

Comment:
It confused me for a moment that the class is called `UnmergeReplacement` regardless of whether the operation is increase or decrease.

Reply:
This is now gone after a discussion on the Snuba PR